### PR TITLE
[Filesystem] improved errors for invalid dump targets

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24097
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/Filesystem/Filesystem.php [hunk 1]
```diff
@@ -532,6 +532,14 @@ public function dumpFile($filename, $content, $mode = 0666)
         $dir = dirname($filename);
 
         if (!is_dir($dir)) {
+            if (is_file($dir)) {
+                throw new IOException(sprintf('The parent path of "%s" is a file.', $dir));
+            }
+
+            if (is_link($dir)) {
+                throw new IOException(sprintf('The parent path of "%s" is a symlink to a not existing directory.', $dir));
```
Comment (javiereguiluz) [inline src/Symfony/Component/Filesystem/Filesystem.php]: Would `nonexistent (dir)` sound better than `not existing (dir)`?

**Comment by robfrawley** [comment]
@xabbuh This pull request works great, and without error when applied to `symfony/symfony:2.7.x-dev` during a `liip/imagine-bundle` test suite run, resolving https://github.com/symfony/symfony/pull/24105#issuecomment-328130035!

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/Filesystem/Tests/FilesystemTest.php [hunk 2]
```diff
@@ -1102,6 +1102,37 @@ public function testDumpKeepsExistingPermissionsWhenOverwritingAnExistingFile()
...
+        $this->filesystem->dumpFile($target, 'baz');
+    }
+
+    /**
+     * @expectedException \Symfony\Component\Filesystem\Exception\IOException
+     * @expectedExceptionMessageRegExp /^The parent path of ".*" is a symlink to a nonexistent directory\.$/
+     */
+    public function testDumpFailsIfTargetIsSymlinkAndSymlinkTargetDoesNotExist()
+    {
+        $this->markAsSkippedIfSymlinkIsMissing();
```
Comment (xabbuh) [inline src/Symfony/Component/Filesystem/Tests/FilesystemTest.php]: We need to check why this does not apply on Appveyor or if `is_link()` works differently on Windows than on Unix. If someone using Windows can help here, that would be really appreciated.

File: src/Symfony/Component/Filesystem/Filesystem.php [hunk 3]
```diff
@@ -525,13 +525,21 @@ public function isAbsolutePath($file)
      * @param null|int $mode     The file mode (octal). If null, file permissions are not modified
      *                           Deprecated since version 2.3.12, to be removed in 3.0.
      *
-     * @throws IOException If the file cannot be written to.
+     * @throws IOException if the file cannot be written to
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Filesystem/Filesystem.php]: I think we disabled the fabbot rule triggering this, can you check?

Comment (fabpot) [reply src/Symfony/Component/Filesystem/Filesystem.php]: No, I've fixed everything in the codebase... using php-cs-fixer.

**Comment by nicolas-grekas** [comment]
Failure on Windows looks related.

Status: needs work

**Comment by fabpot** [comment]
@xabbuh Can you have a look at the tests?

**Comment by fabpot** [comment]
@xabbuh Can you try to finish this one?

**Comment by fabpot** [comment]
soft ping @xabbuh

**Comment by xabbuh** [comment]
I have no idea how to make the test pass on Windows and I don't have access to a Windows machine to debug this properly. I am closing here for now. Anyone wanting to dig into the issue can reuse the changes here. I will gladly help with any questions regard the current state of the changes.

**Pull request closed without merging**